### PR TITLE
[BUGFIX release] Fix uglification introduced bug with super wrapping.

### LINKED ISSUE #12462: Dev/Prod inconsistency for Ember.Object.create().hasOwnProperty('_super')
Hey,

today I ran into an issue with our production builds being broken,  which was ultimately caused by minified builds behaving differently to non-minified builds:
`Ember.Object.create().hasOwnProperty('_super')` is `false` in non-minified builds, but `true` in minified ones.

I've actually verified that building in `production` environment with `minifyJS: { enabled: false }` isn't making `_super` an 'own property'. So I assume it's a minification issue and nothing else that might differ for production and development builds.

We stumbled upon this because we were doing `for in` on an `Ember.Object` and checking the properties with `hasOwnProperty`. (yes, it's fishy. no, i don't think it's a good idea. yes, i've changed the code already, but the inconsisteny *somewhere* remains).

To show this behavior I prepared a demo project here:
https://github.com/dschmidt/ember-dev-prod-inconsistency
(It's just a completely empty project ... but any of your projects should be good)

Just run `ember s` and `ember s --environment=production` and compare the outputs of `Ember.Object.create().hasOwnProperty('_super')`

BR,
Domme

### PR DESCRIPTION
**Pull request by rwjblue**
We check to see if a function contains `_super` / `.call` / `.apply` before attempting to super wrap a given method (this saves quite a bit of extra super wrapping for functions that do not need it).  Unfortunately, a number of platforms that we support do not support calling `func.toString()` so we attempt to detect this and fall back to always super wrap everything mode.

We have roughly this code (from [here](https://github.com/emberjs/ember.js/blob/b4718218dbe5ffe7736c485a594248b20977c621/packages/ember-metal/lib/utils.js#L257-L271)) to detect if we can call `toString()` and get the original source:

``` javascript
  let sourceAvailable = (function() {
    return this;
  }).toString().indexOf('return this;') > -1;
```

This works perfectly for development builds, but unfortunately not when minified.  Take a look at the minified source and you will see why:

``` javascript
var e=function(){return this}.toString().indexOf("return this;")>-1;
```

Note that minifier has stripped the trailing semicolon from the function body we are attempting to check against, but sadly our `indexOf` check still contains the semicolon!

**tldr;** We super wrap every function in uglified builds.

---

This commit removes the trailing semicolon from the source check, and adds a test that should hopefully allow us to detect if this is accidentally reintroduced.

Fixes #12462.

### CODE REVIEW

**Comment by stefanpenner** [comment]
Lgtm

**Comment by ef4** [comment]
LGTM. 

The test will fail on platforms that don't support `Function.toString`, but that is kinda the point of the test: detecting unnecessary super wrapping, which those platforms unfortunately do have.

**Comment by rwjblue** [comment]
We could add a phantomjs detection to https://github.com/emberjs/ember.js/blob/master/packages/ember-metal/lib/environment.js#L27-L28, and wrap the test in a `if (env.isPhantom || env.isChrome || env.isFirefox)`.  This allows the test suite to pass on other browers that may or may not need the "always wrap" fallback path, but ensures we continue to pass on browsers that we know should support `function() { }.toString()`.

Thoughts?

**Comment by rwjblue** [comment]
Added the phantomjs check I mentioned above, the test suite should now continue to pass even on platforms that do not support function source.

**Comment by ef4** [comment]
:+1:

**Comment by dschmidt** [comment]
Awesome, thanks!

**Comment by stefanpenner** [comment]
this is likely going to be a nice little perf post.
